### PR TITLE
Fixes MMIs in borgs dying to ash storms.

### LINKED ISSUE #38713: Ash storms damaging/killing borg MMI's (while inside the borg body).
Issue reported from  Round ID: 90005 (/tg/Station Bagil [ENGLISH] [US-EAST] [100% FREE LAG])

Reporting client version: 512

[Round ID]: # Happened on multiple rounds.

[Reproduction]: # Went mining as a borg. Given that borgs are (from what I can tell) are immune to damage from ash storms, I generally refrained from seeking out shelter when one started up. However, upon removal of my MMI (disassembly by a roboticist), the MMI was no longer able to be put back into my borg body. The MMI's appearance was "dead", but I was able to talk in character for around 30s before I began talking in deadchat. After consulting with an admin, they found that my MMI had received 200 fire damage rapidly which led to my MMI being killed. I've also tested this where I've avoided ash storms and gotten my MMI removed and it does _not_ exhibit this issue.

1) Be a borg and chill in an ash storm. 
2) MMI is removed for any reason.
3) MMI will eventually die from rapid fire damage.

I do not know if positronic brains are affected by this issue. I'm assuming it's a bug because I feel that the borg body should protect the MMI from damage.

### PR DESCRIPTION
**Pull request by zxaber**
Fixes #38713 
Credit is basically Oranges.

:cl: Zxaber
fix: MMI units inside cyborgs are no longer affected by ash storms.
/:cl:

Borgs are supposed to be ash-proof, and they are. Their (organic) MMIs, however, are not, and take burn damage while the borg is in a storm. This does not result is issues until the MMI is removed from the borg, at which point it's basically dead sans being able to talk for about thirty seconds. You cannot place the brain into a new shell at any time until you clone it. 

This fixes that. Thanks to Oranges, normal unborged MMIs (and brains) are still affected by storms. If and when I can figure out the better way to fix this (remaking weather immunity), I'll change it later, but this at least fixes the bug in the meantime.

### CODE REVIEW

**Comment by ExcessiveUseOfCobblestone** [comment]
I don't think this works?

**Review by ExcessiveUseOfCobblestone — CHANGES_REQUESTED** [review]

**Comment by ExcessiveUseOfCobblestone** [comment]
Just out of curiosity how are borgs getting ash-immunity prior to this change?

**Comment by zxaber** [comment]
Far as I can tell, it's done in modules/mob/living/silicon/silicon.dm; line eleven is `weather_immunities = list("ash")`. This change just gives the brain in the MMI (and technically, any living entity inside another living entity) the ability to inherit it's ash-proof status.

Technically speaking, the cyborg itself is currently ash-proof, but the brain inside the MMI is not. So the brain racks up the 200 burn damage, but no effect happens from this until the MMI is removed from the borg.

(My guess is that the game considers the shell as the ghost/player's body until the shell is removed, which transfers the ghost/player to the brain, which then suddenly has to take the burn damage into consideration.)

**Comment by ExcessiveUseOfCobblestone** [comment]
My question is where is it checking for `weather_immunities` being ash until this PR?

File: code/datums/weather/weather_types/ash_storm.dm [hunk 1]
```diff
@@ -83,7 +83,11 @@
 			var/thermal_protection = H.get_thermal_protection()
 			if(thermal_protection >= FIRE_IMMUNITY_SUIT_MAX_TEMP_PROTECT)
 				return TRUE
-		L = L.loc //Matryoshka check
+		if(isliving(L))// if we're a non immune mob inside an immune mob we have to reconsider if that mob is immune to protect ourselves
+			var/mob/living/the_mob = L
+			if("ash" in the_mob.weather_immunities)
+				return TRUE
```
[the remark this answers is not in the document]
Comment (zxaber) [reply code/datums/weather/weather_types/ash_storm.dm]: To be sure, you mean move the human if-block to be nested into the isliving if-block, right?

**Review by zxaber — COMMENTED** [review]
Oh, er, hmm.

In weather.dm, there is

107: `/datum/weather/proc/can_weather_act(mob/living/L) //Can this weather impact a mob`
108: `var/turf/mob_turf = get_turf(L)`
..
111: `	if(immunity_type in L.weather_immunities)`
112: ` return`

Up on line 22 of this ash_storm.dm file, it defines the immunity as ash;
22: `	immunity_type = "ash"`

And of course "ash" is added to that list in the silicon.dm file.

It seems weather.dm originally starts up the storm, then pulls ash-storm-specific info from ash_storm.dm.

File: code/datums/weather/weather_types/ash_storm.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (ExcessiveUseOfCobblestone) [reply code/datums/weather/weather_types/ash_storm.dm]: yep, if it's not living it shouldn't bother checking if it's human.

**Comment by ExcessiveUseOfCobblestone** [comment]
Actually wait this whole issue stems from it having it's own snowflake proc rather than using `can_weather_act`.

**Comment by ExcessiveUseOfCobblestone** [comment]
How about this?

```dm
/datum/weather/ash_storm/proc/is_ash_immune(atom/L)
    while(L && !isturf(L))
        if(isliving(L))
            if(!can_weather_act(L)) //you're immune to ash-storms, harry!
                return TRUE
        if(ismecha(L))
            return TRUE
        L = L.loc // Pushing_Upwards.wav
    return FALSE

/datum/weather/ash_storm/can_weather_act(mob/living/L)
    if(ishuman(L))
        var/mob/living/carbon/human/H = L
        if(H.get_thermal_protection() >= FIRE_IMMUNITY_SUIT_MAX_TEMP_PROTECT)
            return FALSE
    ..() //weather_immunities etc etc
```

**Comment by zxaber** [comment]
`can_heather_act` is in weather.dm, not ash_storm.dm, Is it going to work if I put it in ash_storm.dm too?

**Comment by zxaber** [comment]
Seems to work, updated

**Comment by ExcessiveUseOfCobblestone** [comment]
I'm running a faith build of course I know the miracles!

File: code/datums/weather/weather.dm [hunk 2]
```diff
@@ -105,6 +105,10 @@
 	update_areas()
 
 /datum/weather/proc/can_weather_act(mob/living/L) //Can this weather impact a mob?
+	if(ishuman(L))
+		var/mob/living/carbon/human/H = L
+		if(H.get_thermal_protection() >= FIRE_IMMUNITY_SUIT_MAX_TEMP_PROTECT)
```
Comment (vuonojenmustaturska) [inline code/datums/weather/weather.dm]: this belongs in ash storm weather, not `/datum/weather`

**Review by vuonojenmustaturska — DISMISSED** [review]

File: code/datums/weather/weather.dm (hunk 2 above)
Comment (zxaber) [reply code/datums/weather/weather.dm]: `can_weather_act` existed in weather.dm before I made any changes. If I make a separate `can_weather_act` in ash_storm.dm, is it going to cause issues?

Comment (ExcessiveUseOfCobblestone) [reply code/datums/weather/weather.dm]: no because (using the code i wrote earlier) it calls inher.

Comment (zxaber) [reply code/datums/weather/weather.dm]: If I move the whole thing into Ash_storm, it's going to break other storms, right? So what should I do?

Comment (zxaber) [reply code/datums/weather/weather.dm]: So I reverted the weather.dm file and added `can_weather_act` in ash_storm.dm as well, and now the storms damage nothing whatsoever. weather.dm had a `can_weather_act` section too, so I'm guessing it's a conflict?

Comment (ExcessiveUseOfCobblestone) [reply code/datums/weather/weather.dm]: you do not touch this proc and copy what I put in 

```DM
/datum/weather/ash_storm/can_weather_act(mob/living/L)
    if(ishuman(L))
        var/mob/living/carbon/human/H = L
        if(H.get_thermal_protection() >= FIRE_IMMUNITY_SUIT_MAX_TEMP_PROTECT)
            return FALSE
    ..() //weather_immunities etc etc
```

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**